source/org/zfin/construct/presentation/ConstructComponentService.java

```diff
@@ -308,5 +314,127 @@ private static void validateConstructNameOrThrowException(String constructName)
             throw new InvalidConstructNameException(getValidationErrorMessageForConstructName(constructName).orElse("Invalid construct name"));
         }
     }
+
+    /**
+     * Get the ConstructName object from the construct ID
+     * It retrieves the ConstructComponent objects from the database and parses them into a ConstructName object
+     * @param constructID
+     * @return
+     */
+    public static ConstructName getExistingConstructName(String constructID) {
+        List<ConstructComponent> components = cr.getConstructComponentsByConstructZdbId(constructID);
+        Iterator<ConstructComponent> componentIterator = components.iterator();
+        ConstructName constructName = new ConstructName();
+
+        //Set the TYPE
+        ConstructComponent currentComponent = componentIterator.next();
+        constructName.setTypeByAbbreviation(currentComponent.getComponentValue());
+
+        //Set the PREFIX if it exists
+        currentComponent = componentIterator.next();
+        if (currentComponent.getComponentValue().equals("(")) {
+            constructName.setPrefix("");
+        } else {
+            constructName.setPrefix(currentComponent.getComponentValue());
+        }
+
+        //Set the Cassettes
+        Cassette currentCassette = new Cassette();
+        Promoter currentPromoter = new Promoter();
+        Coding currentCoding = new Coding();
+        int cassetteNumber = 1;
+        int componentNumber = 2;
+        String parsingCassette = "promoter";
+        while (componentIterator.hasNext()) {
+            componentNumber++;
+            currentComponent = componentIterator.next();
+            if(cassetteNumber != currentComponent.getComponentCassetteNum()) {
+                currentCassette.setPromoter(currentPromoter);
+                currentCassette.setCoding(currentCoding);
+                constructName.addCassette(currentCassette);
+                currentCassette = new Cassette();
+                cassetteNumber++;
+                currentPromoter = new Promoter();
+                currentCoding = new Coding();
+                parsingCassette = "promoter";
+            }
+            if (currentComponent.getComponentValue().equals(COMPONENT_SEPARATOR)) {
+                continue;
+            } else if (currentComponent.getComponentValue().equals(")") && !componentIterator.hasNext()) {
+                continue;
+            } else if (currentComponent.getComponentCategory().equals("promoter component")) {
+                currentPromoter.addPromoterPart(currentComponent.getComponentValue());
+            } else if (currentComponent.getComponentCategory().equals("coding component")) {
+                parsingCassette = "coding";

```

Comment:
Can that we re-written with a switch statement?

Reply:
Good idea. I'll change it

Reply:
Updated